parley/src/shape.rs

```diff
@@ -215,10 +196,194 @@ pub(crate) fn shape_text<'a, B: Brush>(
+    let mut font_selector =
+        FontSelector::new(fq, rcx, styles, first_style_index, item.script, item.locale);
+
+    // Parse text into clusters of the current item
+    let tokens =
+        item_text
+            .char_indices()
+            .zip(item_infos)
+            .map(|((offset, ch), (info, style_index))| Token {
+                ch,
+                offset: (text_range.start + offset) as u32,
+                len: ch.len_utf8() as u8,
+                info: *info,
+                data: *style_index as u32,
+            });
+
+    let mut parser = swash::text::cluster::Parser::new(item.script, tokens);
+    let mut cluster = CharCluster::new();
+
+    // Reimplement swash's shape_clusters algorithm - but only for current item
+    if !parser.next(&mut cluster) {
+        return; // No clusters to process
+    }
+
+    let mut current_font = font_selector.select_font(&mut cluster);
+
+    // Main segmentation loop (based on swash shape_clusters) - only within current item
+    while let Some(font) = current_font.take() {
+        // Collect all clusters for this font segment
+        let segment_start_offset = cluster.range().start as usize - text_range.start;
+        let mut segment_end_offset = cluster.range().end as usize - text_range.start;
+
+        loop {
+            if !parser.next(&mut cluster) {
+                // End of current item - process final segment
+                break;
+            }
+
+            if let Some(next_font) = font_selector.select_font(&mut cluster) {
+                if next_font != font {
+                    current_font = Some(next_font);
+                    break;
+                } else {
+                    // Same font - add to current segment
+                    segment_end_offset = cluster.range().end as usize - text_range.start;
+                }
+            } else {
+                // No font found - skip this cluster
+                if !parser.next(&mut cluster) {
+                    break;
+                }
+            }
+        }
+
+        // Shape this font segment with harfrust
+        let segment_text = &item_text[segment_start_offset..segment_end_offset];
+        // Shape the entire segment text including newlines
+        // The line breaking algorithm will handle newlines automatically
+
+        // TODO: How do we want to handle errors like this?
+        let font_ref =
+            harfrust::FontRef::from_index(font.font.blob.as_ref(), font.font.index).unwrap();
+
+        // Create harfrust shaper
+        // TODO: cache this upstream?
+        let shaper_data = harfrust::ShaperData::new(&font_ref);
+        let instance = harfrust::ShaperInstance::from_variations(
+            &font_ref,
+            variations_iter(&font.font.synthesis, rcx.variations(item.variations)),
+        );
+        // TODO: Don't create a new shaper for each segment.
+        let harf_shaper = shaper_data
+            .shaper(&font_ref)
```

Comment:
If we can easily retain this then let's do it. Otherwise, it's not _that_ expensive to construct.